.github/workflows/ci.yml

```diff
@@ -11,37 +11,40 @@ on:
     types:
       released
 
+permissions:
+  contents: read
+
 jobs:
   test:
     runs-on: envoy-x64-small
     if: github.repository_owner == 'envoyproxy'
     steps:
-    - uses: actions/checkout@v4
-    - uses: actions/setup-python@v5
+    - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+    - uses: actions/setup-python@0a5c61591373683505ea898e09a3ea4f39ef2b9c # v5.0.0
       with:
         python-version: "3.9"
     - name: Cache pants
-      uses: actions/cache@v3
+      uses: actions/cache@704facf57e6136b1bc63b828d79edcd491f0ee84 # v3.3.2
       with:
         key: ${{ runner.os }}-${{ hashFiles('pants*toml') }}
         path: ~/.cache/pants/setup
     - name: Run pants test
       run: "./pants --colors test ::"
     - name: Archive code coverage results
-      uses: actions/upload-artifact@v3
+      uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
```

Comment:
im really in 2 minds about using hashes for these - obv its a lot more secure and preferable from that pov but its totally un-user friendly

this one is a good example - v4 of download/upload has some breaking changes which i need to track down - ive already reverted them once downstream - its impossible to track that with shas

i dont understand why they dont allow specifying both version and sha as you can with Docker

Reply:
ah wait - i see the versions are commented - i guess that can work - does dependabot respect that?

Reply:
Yes dependabot handles this greatly